Fix panic on init for go-getter modules w version

sl1pm4t opened the pull request:
Fixes #21328

If the `version` attribute is used on modules whose source will be retrieve with `go-getter` (not Registry or local relative file paths), then Terraform will panic and crash.

This PR fixes the panic by preventing the use of the `version` attribute for module sources that would be retrieved with go-getter. I'm not sure if there are any cases where the version constraint should be allowed in that scenario?

apparentlymart commented on internal/initwd/test-fixtures/invalid-version-constraint/root.tf:
```diff
@@ -0,0 +1,7 @@
+# This fixture depends on a github repo at:
+#     https://github.com/hashicorp/terraform-aws-module-installer-acctest
```
If this were true then we'd need to put the `TF_ACC=1` guard at the start of the test to ensure it isn't included in a normal test run (as you can see in e.g. `TestLoaderInstallModules_registry`.

However, I think this isn't actually true in practice, right? Because the presence of the version constraint will block Terraform from ever attempting to fetch this repository. If that's so, I think it'd be helpful to future maintainers for the comment to explain that the address here is never actually fetched because of the error.

sl1pm4t commented:
Hi @apparentlymart 

A two part fix based on your comment:

> Because the presence of the version constraint will block Terraform from ever attempting to fetch this repository.

1. I updated the test-fixtures comment as you suggested.
2. However I found despite the error, the module installer was still retrieving the module from GitHub. I've now added a return in `installGoGetterModule()` to short circuit if the *invalid version constraint* error is detected. The same should probably be done in `installLocalModule()`.

In addition, I improved the unit test to check the error message is the one we expect.

apparentlymart commented:
Thanks, @sl1pm4t! This is looking great. I'm going to merge it now.
